unixfs/src/dir/builder.rs

```diff
@@ -0,0 +1,834 @@
+                    old_depth,
+                    depth
+                );
+            }
+        }
+    }
+
+    debug_assert!(*old_depth <= depth);
+
+    if let Some(name) = name {
+        if !full_path.is_empty() {
+            full_path.push_str("/");
+        }
+        full_path.push_str(name);
+        *old_depth += 1;
+    }
+
+    assert_eq!(*old_depth, depth);
+}
+
+pub struct PostOrderIterator<'a> {
+    full_path: &'a mut String,
+    old_depth: usize,
+    block_buffer: &'a mut Vec<u8>,
+    // our stack of pending work
+    pending: Vec<Visited>,
+    // "communication channel" from nested entries back to their parents
+    persisted_cids: HashMap<Option<u64>, BTreeMap<String, Leaf>>,
+    reused_children: Vec<Visited>,
+    cid: Option<Cid>,
+    // from TreeOptions
+    wrap_in_directory: bool,
+}
+
+fn identity_cid(number: usize) -> Cid {
+    use multihash::{Multihash, Sha2_256};
+
+    let mh = Sha2_256::digest(&number.to_le_bytes());
+    Cid::new_v0(mh).unwrap()
+}
+
+#[derive(Debug)]
+pub enum TreeConstructionFailed {
+    // TODO: at least any quick_protobuf errors here?
+}
+
+impl fmt::Display for TreeConstructionFailed {
+    fn fmt(&self, _fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        todo!()
+    }
+}
+
+impl std::error::Error for TreeConstructionFailed {}
+
+impl<'a> PostOrderIterator<'a> {
+    fn render_directory(
+        links: &BTreeMap<String, Leaf>,
+        buffer: &mut Vec<u8>,
+    ) -> Result<Leaf, TreeConstructionFailed> {
+        use crate::pb::{FlatUnixFs, PBLink, UnixFs, UnixFsType};
+        use quick_protobuf::{BytesWriter, MessageWrite, Writer};
+        use sha2::{Digest, Sha256};
+        use std::borrow::Cow;
+
+        // TODO: this could quite easily be made so that the links are read from the btreemap for
+        // calculating the size and rendering
+        let mut combined_from_links = 0;
+
+        let flat = FlatUnixFs {
+            links: links
+                .iter() // .drain() would be the most reasonable
+                .inspect(|(_, Leaf { total_size, .. })| combined_from_links += total_size)
+                .map(|(name, Leaf { link, total_size })| PBLink {
+                    Hash: Some(link.to_bytes().into()),
+                    Name: Some(Cow::Borrowed(name.as_str())),
+                    Tsize: Some(*total_size),
+                })
+                .collect::<Vec<_>>(),
+            data: UnixFs {
+                Type: UnixFsType::Directory,
+                Data: None,
+                ..Default::default()
+            },
+        };
+
+        let size = flat.get_size();
+
+        // FIXME: we shouldn't be creating too large structures (bitswap block size limit!)
+        // FIXME: changing this to autosharding is going to take some thinking
+
+        buffer.clear();
+        let cap = buffer.capacity();
+
+        if let Some(additional) = size.checked_sub(cap) {
+            buffer.reserve(additional);
+        }
+
+        // argh
```

Comment:
```suggestion
        // FIXME: argh
```

:)

Reply:
basically just never clearing should work, fill only for larger entries, slicing should probably handle it.